File: core/context.lua
context = { }
context.__index = context

function context.new()
	local self = { }
	setmetatable(self, context)

	self._input = input.new(self)
	self._painter = painter.new(self)

	self._isDebug = false
	self._state = { }
	self._nextState = { }

	return self
end

function context:setNextWindowNoDrag(isNoDrag)
	self._state.isNoDrag = isNoDrag
end

function context:isWindowNoDrag()
	return self._state.isNoDrag
end

function context:beginWindow(x, y)
	self._input:beginWindow()
	self._painter:beginWindow(x, y)
end

function context:endWindow()
	local x, y = self._painter:endWindow()

	self._input:endWindow()

	self._state = { }

	return x, y
end

function context:isWidgetHovered()
	return self._input:isRectHovered(self._painter:getWidgetX(), self._painter:getWidgetY(), self._painter:getWidgetWidth(), self._painter:getWidgetHeight())
end

function context:isWidgetClicked()
	return self._input:isMousePressed() and self:isWidgetHovered()
end

function context:beginDraw(w, h)
	self._painter:beginDraw(w, h)
end

function context:endDraw()
	self._painter:endDraw()

	self._nextState = { }
end

function context:setDebugEnabled(enabled)
	self._isDebug = enabled
end

function context:isDebugEnabled()
	return self._isDebug
end

function context:__setTextEntry(state, entry, ...)
	state.textEntry = entry
	state.textComponents = { ... }
end

function context:setNextTextEntry(entry, ...)
	self:__setTextEntry(self._nextState, entry, ...)
end

function context:pushTextEntry(entry, ...)
	self:__setTextEntry(self._state, entry, ...)
end

function context:popTextEntry()
	self._state.textEntry = nil
	self._state.textComponents = nil
end

function context:getTextEntry()
	return self._nextState.textEntry or self._state.textEntry
end

function context:getTextComponents()
	return self._nextState.textComponents or self._state.textComponents
end

function context:__setWidgetWidth(state, w)
	state.widgetWidth = w
end

function context:setNextWidgetWidth(w)
	self:__setWidgetWidth(self._nextState, w)
end

function context:pushWidgetWidth(w)
	self:__setWidgetWidth(self._state, w)
end

function context:popWidgetWidth()
	self._state.widgetWidth = nil
end

function context:getWidgetWidth()
	return self._nextState.widgetWidth or self._state.widgetWidth
end

function context:getInput()
	return self._input
end

function context:getPainter()
	return self._painter
end

File: core/context.js
function setTextEntry(state, entry, ...args) {
	state.textEntry = entry;
	state.textComponents = args;
}

function setWidgetWidth(state, w) {
	state.widgetWidth = w;
}

class Context {
	#input;
	#painter;
	#isDebug;
	#state;
	#nextState;

	constructor() {
		this.#input = new Input();
		this.#painter = new Painter(this);

		this.#isDebug = false;
		this.#state = {};
		this.#nextState = {};
	}

	setNextWindowNoDrag(isNoDrag) {
		this.#state.isNoDrag = isNoDrag;
	}

	isWindowNoDrag() {
		return this.#state.isNoDrag;
	}

	beginWindow(windowPos) {
		this.#input.beginWindow();
		this.#painter.beginWindow(windowPos);
	}

	endWindow() {
		const windowPos = this.#painter.endWindow();

		this.#input.endWindow();

		this.#state = {};

		return windowPos;
	}

	isWidgetHovered() {
		return this.#input.isRectHovered(
			this.#painter.getWidgetX(),
			this.#painter.getWidgetY(),
			this.#painter.getWidgetWidth(),
			this.#painter.getWidgetHeight()
		);
	}

	isWidgetClicked() {
		return this.#input.isMousePressed() && this.isWidgetHovered();
	}

	beginDraw(w, h) {
		this.#painter.beginDraw(w, h);
	}

	endDraw() {
		this.#painter.endDraw();

		this.#nextState = {};
	}

	setDebugEnabled(enabled) {
		this.#isDebug = enabled;
	}

	isDebugEnabled() {
		return this.#isDebug;
	}

	setNextTextEntry(entry, ...args) {
		setTextEntry(this.#nextState, entry, ...args);
	}

	pushTextEntry(entry, ...args) {
		setTextEntry(this.#state, entry, ...args);
	}

	popTextEntry() {
		this.#state.textEntry = null;
		this.#state.textComponents = null;
	}

	getTextEntry() {
		return this.#nextState.textEntry || this.#state.textEntry;
	}

	getTextComponents() {
		return this.#nextState.textComponents || this.#state.textComponents;
	}

	setNextWidgetWidth(w) {
		setWidgetWidth(this.#nextState, w);
	}

	pushWidgetWidth(w) {
		setWidgetWidth(this.#state, w);
	}

	popWidgetWidth() {
		this.#state.widgetWidth = nil;
	}

	getWidgetWidth() {
		return this.#nextState.widgetWidth || this.#state.widgetWidth;
	}

	getInput() {
		return this.#input;
	}

	getPainter() {
		return this.#painter;
	}
}
